drivers/entropy/entropy_wdt.c

```diff
@@ -0,0 +1,267 @@
+ * with the number of CPU cycles at the time the system timer expires,
+ * a bit of entropy is generated and accumulated into a sample byte.
+ *
+ * Every sample that is generated by the race is then stored in a pool.
+ * The race is stopped when this pool is full, to reduce power
+ * consumption due to increased wake ups to service the interrupts; it
+ * starts again as soon as bytes are consumed from the pool.
+ *
+ * Since only the least 8 significant bits of the difference is
+ * considered when accumulating into the sample byte, it's important to
+ * keep the CPU cycles difference as close as possible.  Each time a
+ * sample contains 8 bytes, the system timer is restarted in an effort
+ * to reduce the drift between the two timers.
+ *
+ * While this effectively solves the issue of not having enough entropy
+ * on devices without a hardware random number generator, it also makes
+ * use of the watchdog timer that the application developer might want
+ * to consider using.
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <watchdog.h>
+
+#include <errno.h>
+
+/* 2^16 cycles for the watchdog is 2.048ms; if changing one value, the
+ * other one needs to be changed so it's as close as possible.
+ *
+ * Keep in mind that this generates close to 1bit of entropy every time
+ * these timers expire, so to generate a full byte, it's going to take
+ * about 16.384ms.  To generate enough entropy to initialize a PRNG
+ * such as rand32_xoroshiro128.c, this will take about 65.536ms.
+ */
+#define WATCHDOG_TIMEOUT WDT_2_16_CYCLES
+#define TIMER_TIMEOUT K_MSEC(2)
+
+struct entropy_wdt_data {
+	struct device *wdt;
+	struct k_timer timer;
+
+	/* Struct protected by `ewd_entropy_sem`, defined below. */
+	struct {
+		u8_t buffer[16];
+		u8_t first;
+		u8_t last;
+		u8_t population;
+
+		bool need_restart;
+	} entropy;
+
+	/* Struct protected by `ewd_sample_sem`, defined below. */
+	struct {
+		u32_t cycle_wdt_timer;
+
+		u8_t current_byte;
+		u8_t bits;
+
+		bool has_wdt_sample;
+	} sample;
+};
+
+static struct entropy_wdt_data ewd;
+static K_SEM_DEFINE(ewd_entropy_sem, 1, 1);
+static K_SEM_DEFINE(ewd_sample_sem, 1, 1);
+
+static void entropy_wdt_stop_timers(void)
+{
+	wdt_disable(ewd.wdt);
+	k_timer_stop(&ewd.timer);
+}
+
+static void entropy_wdt_start_timers(void)
+{
+	wdt_enable(ewd.wdt);
+	k_timer_start(&ewd.timer, TIMER_TIMEOUT, TIMER_TIMEOUT);
+}
+
+static bool entropy_wdt_put_byte(void)
+{
+	bool stopped_timer;
+	int ret;
+
+	ret = k_sem_take(&ewd_entropy_sem, K_FOREVER);
+	if (ret < 0) {
+		return false;
+	}
+
+	ewd.entropy.buffer[ewd.entropy.last] = ewd.sample.current_byte;
+	ewd.entropy.last = (ewd.entropy.last + 1) & 15;
+	ewd.entropy.population = (entropy.population + 1) & 15;
+
+	if (!ewd.entropy.population) {
+		/* The entropy pool is full.  Disable the timers
+		 * temporarily to reduce power consumption.
+		 *
+		 * FIXME: consider not turning them off, but slowing
```

Comment:
What are the tradeoffs here? Is there a slow setup? At least for mbed TLS use, it is going to ask for a bunch of entropy at start, and then only after a period of time when it decides it needs to stir more in. With typical TLS use, this would happen on the order of many days, even with heavy network use.

Reply:
If the timers keep running when the entropy ring buffer gets full, bytes will just be overwritten.  I decided to stop the timers here because not running any code between seeds of a PRNG seems like a good idea to me, specially since this might influence power consumption.